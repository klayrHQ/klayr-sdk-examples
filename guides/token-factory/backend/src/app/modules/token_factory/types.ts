import { JSONObject } from 'klayr-sdk';

export type TokenID = Buffer;

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: bigint;
	minAmountToMint: bigint;
	maxAmountToMint: bigint;
	chainID: Buffer;
	createTokenFee: bigint;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
