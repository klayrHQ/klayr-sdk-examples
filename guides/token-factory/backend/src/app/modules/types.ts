import { JSONObject } from 'lisk-sdk';

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: number;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
