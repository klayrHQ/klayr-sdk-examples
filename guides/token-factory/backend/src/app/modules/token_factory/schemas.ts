export const mintSchema = {
	$id: 'token_factory/mintParams',
	title: 'Mints extra tokens to a recipient',
	type: 'object',
	properties: {
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		recipient: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'klayr32',
		},
	},
	required: ['amount', 'recipient'],
};

export const createTokenSchema = {
	$id: 'token_factory/createToken-params',
	title: 'CreateTokenCommand transaction parameter for the Token Factory module',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export const configSchema = {
	$id: '/token_factory/config',
	type: 'object',
	properties: {
		maxNameLength: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		maxSymbolLength: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		maxTotalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		minAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 4,
		},
		maxAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 5,
		},
		createTokenFee: {
			dataType: 'uint64',
			fieldNumber: 6,
		},
	},
	required: [
		'maxNameLength',
		'maxSymbolLength',
		'maxTotalSupply',
		'minAmountToMint',
		'maxAmountToMint',
		'createTokenFee',
	],
};
