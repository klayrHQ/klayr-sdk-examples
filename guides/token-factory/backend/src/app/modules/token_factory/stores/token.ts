import { BaseStore } from 'klayr-sdk';

export interface TokenStoreData {
	name: string;
	symbol: string;
	totalSupply: bigint;
}

export const tokenStoreSchema = {
	$id: '/token_factory/token',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export class TokenStore extends BaseStore<TokenStoreData> {
	public schema = tokenStoreSchema;
}
