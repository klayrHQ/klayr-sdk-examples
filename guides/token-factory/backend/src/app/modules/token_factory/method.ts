import { BaseMethod, CommandVerifyContext } from 'klayr-sdk';
import { OwnerStore, OwnerStoreData } from './stores/owner';

export class TokenFactoryMethod extends BaseMethod {
	public async getOwnerOfToken<T>(
		context: CommandVerifyContext<T>,
		tokenID: Buffer,
	): Promise<OwnerStoreData | undefined> {
		try {
			const ownerStore = this.stores.get(OwnerStore);
			return await ownerStore.get(context, tokenID);
		} catch {
			return undefined;
		}
	}
}
