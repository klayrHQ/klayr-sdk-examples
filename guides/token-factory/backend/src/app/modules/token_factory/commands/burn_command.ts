/* eslint-disable */
import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
	TokenMethod,
} from 'klayr-sdk';
import { burnSchema } from '../schemas';
import { failWithLog } from '../utils';
import { TokenID } from '../types';
import { TokenStore } from '../stores/token';
import { TokenFactoryMethod } from '../method';

export interface BurnParams {
	tokenID: TokenID;
	amount: bigint;
}

export class BurnCommand extends BaseCommand {
	private _minAmountToBurn!: bigint;
	private _tokenFactoryMethod!: TokenFactoryMethod;
	private _tokenMethod!: TokenMethod;

	public schema = burnSchema;

	public addDependencies(args: {
		tokenFactoryMethod: TokenFactoryMethod;
		tokenMethod: TokenMethod;
	}) {
		this._tokenFactoryMethod = args.tokenFactoryMethod;
		this._tokenMethod = args.tokenMethod;
	}

	public async init(args: { minAmountToBurn: bigint }): Promise<void> {
		this._minAmountToBurn = args.minAmountToBurn;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<BurnParams>): Promise<VerificationResult> {
		const { tokenID, amount } = context.params;

		const isAmountValid = await this.isAmountValid(context, amount, tokenID);
		if (!isAmountValid) {
			return failWithLog<BurnParams>(
				context,
				`Amount can not be lower than ${this._minAmountToBurn} or greater than balance`,
			);
		}

		const owner = await this._tokenFactoryMethod.getOwnerOfToken<BurnParams>(context, tokenID);
		if (!owner) {
			return failWithLog<BurnParams>(context, `Invalid tokenID: ${tokenID}`);
		}

		if (!owner.address.equals(context.transaction.senderAddress)) {
			return failWithLog<BurnParams>(context, `Sender is not the token creator`);
		}

		return { status: VerifyStatus.OK };
	}

	public async execute(context: CommandExecuteContext<BurnParams>): Promise<void> {
		const { tokenID, amount } = context.params;
		const tokenStore = this.stores.get(TokenStore);

		await this._tokenMethod.burn(
			context.getMethodContext(),
			context.transaction.senderAddress,
			Buffer.from(tokenID),
			BigInt(amount),
		);

		const token = await tokenStore.get(context, tokenID);
		token.totalSupply -= BigInt(amount);
		await tokenStore.set(context, tokenID, token);
	}

	private async isAmountValid(
		context: CommandVerifyContext<BurnParams>,
		amount: bigint,
		tokenID: Buffer,
	): Promise<boolean> {
		const availabeBalance = await this._tokenMethod.getAvailableBalance(
			context,
			context.transaction.senderAddress,
			tokenID,
		);

		return amount >= this._minAmountToBurn && amount <= availabeBalance;
	}
}
