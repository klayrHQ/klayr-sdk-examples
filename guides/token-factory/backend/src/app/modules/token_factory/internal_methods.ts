import { BaseMethod, CommandExecuteContext } from 'klayr-sdk';
import { OwnerStore } from './stores/owner';
import { CounterStore, CounterStoreData, counterKey } from './stores/counter';
import { TokenStore } from './stores/token';
import { CreateTokenParams } from './commands/create_token_command';

export class InternalMethod extends BaseMethod {
	private _chainID!: Buffer;

	public async init(chainID: Buffer): Promise<void> {
		this._chainID = chainID;
	}

	public async getNextTokenIDAndSetStores(
		context: CommandExecuteContext<CreateTokenParams>,
		senderAddress: Buffer,
	) {
		const tokenStore = this.stores.get(TokenStore);
		const counterStore = this.stores.get(CounterStore);
		const ownerStore = this.stores.get(OwnerStore);

		let tokenIdCounter: CounterStoreData = await counterStore
			.get(context, counterKey)
			.catch(() => ({ counter: 0 }));
		tokenIdCounter.counter++;

		const newLocalIDBuffer = Buffer.alloc(4);
		newLocalIDBuffer.writeUInt32BE(tokenIdCounter.counter);
		const newTokenIDBuffer = Buffer.concat([this._chainID, newLocalIDBuffer]);

		await Promise.all([
			counterStore.set(context, counterKey, tokenIdCounter),
			ownerStore.set(context, newTokenIDBuffer, { address: senderAddress }),
			tokenStore.set(context, newTokenIDBuffer, {
				name: context.params.name,
				symbol: context.params.symbol,
				totalSupply: BigInt(context.params.totalSupply),
			}),
		]);

		return newTokenIDBuffer;
	}
}
