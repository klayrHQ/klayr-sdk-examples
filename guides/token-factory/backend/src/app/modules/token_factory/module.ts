/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import { validator } from '@klayr/validator';
import {
	BaseModule,
	FeeMethod,
	ModuleInitArgs,
	ModuleMetadata,
	TokenMethod,
	utils,
} from 'klayr-sdk';
import { BatchTransferCommand } from './commands/batch_transfer_command';
import { ModuleConfig, ModuleConfigJSON } from './types';
import { BurnCommand } from './commands/burn_command';
import { CreateTokenCommand } from './commands/create_token_command';
import { MintCommand } from './commands/mint_command';
import { TokenFactoryEndpoint } from './endpoint';
import { NewTokenEvent } from './events/new_token';
import { InternalMethod } from './internal_methods';
import { TokenFactoryMethod } from './method';
import { configSchema } from './schemas';
import { CounterStore } from './stores/counter';
import { OwnerStore } from './stores/owner';
import { TokenStore } from './stores/token';
import { getModuleConfig } from './utils';

export const defaultConfig = {
	maxNameLength: 30,
	maxSymbolLength: 5,
	maxTotalSupply: BigInt(1e18), // max: 9223372036854775807 = 9e18
	minAmountToMint: BigInt(1000),
	maxAmountToMint: BigInt(1e6) * BigInt(1e8),
	createTokenFee: BigInt(100_000),
	minAmountToBurn: BigInt(1000),
};

export class TokenFactoryModule extends BaseModule {
	private readonly _internalMethod = new InternalMethod(this.stores, this.events);

	private _createTokenCommand = new CreateTokenCommand(this.stores, this.events);
	private _batchTransferCommand = new BatchTransferCommand(this.stores, this.events);
	private _mintCommand = new MintCommand(this.stores, this.events);
	private _burnCommand = new BurnCommand(this.stores, this.events);

	private _moduleConfig!: ModuleConfig;
	private _tokenMethod!: TokenMethod;
	private _feeMethod!: FeeMethod;

	public endpoint = new TokenFactoryEndpoint(this.stores, this.offchainStores);
	public tokenFactoryMethod = new TokenFactoryMethod(this.stores, this.events);
	public commands = [
		this._createTokenCommand,
		this._mintCommand,
		this._burnCommand,
		this._batchTransferCommand,
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(TokenStore, new TokenStore(this.name, 0));
		this.stores.register(CounterStore, new CounterStore(this.name, 1));
		this.stores.register(OwnerStore, new OwnerStore(this.name, 2));

		this.events.register(NewTokenEvent, new NewTokenEvent(this.name));
	}

	public addDependencies(tokenMethod: TokenMethod, feeMethod: FeeMethod) {
		this._tokenMethod = tokenMethod;
		this._feeMethod = feeMethod;
		this._createTokenCommand.addDependencies({
			internalMethod: this._internalMethod,
			tokenMethod: this._tokenMethod,
			feeMethod: this._feeMethod,
		});
		this._mintCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._burnCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._batchTransferCommand.addDependencies({
			tokenMethod: this._tokenMethod,
		});
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			commands: this.commands.map(command => ({
				name: command.name,
				params: command.schema,
			})),
			events: this.events.values().map(v => ({
				name: v.name,
				data: v.schema,
			})),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	public async init(args: ModuleInitArgs): Promise<void> {
		// Get the module config defined in the config.json of the node
		const { moduleConfig, genesisConfig } = args;
		// Overwrite the default module config with values from config.json, if set
		const config = utils.objects.mergeDeep({}, defaultConfig, moduleConfig) as ModuleConfigJSON;
		// Validate the config with the config schema

		validator.validate<ModuleConfigJSON>(configSchema, config);

		this._moduleConfig = getModuleConfig(config, genesisConfig);
		this._internalMethod.init(this._moduleConfig.chainID);
		this._createTokenCommand.init(this._moduleConfig).catch(err => {
			console.log('Error: ', err);
		});
		this._mintCommand
			.init({
				minAmountToMint: this._moduleConfig.minAmountToMint,
				maxAmountToMint: this._moduleConfig.maxAmountToMint,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
		this._burnCommand
			.init({
				minAmountToBurn: this._moduleConfig.minAmountToBurn,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
	}

	// public async insertAssets(_context: InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
	// }

	// public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {

	// }
}
