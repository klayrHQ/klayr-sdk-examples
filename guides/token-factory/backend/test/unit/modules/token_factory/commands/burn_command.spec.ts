/* eslint-disable */

import { BurnCommand, BurnParams } from '@app/modules/token_factory/commands/burn_command';
import {
	CreateTokenCommand,
	CreateTokenParams,
} from '@app/modules/token_factory/commands/create_token_command';
import { InternalMethod } from '@app/modules/token_factory/internal_methods';
import { TokenFactoryMethod } from '@app/modules/token_factory/method';
import { TokenFactoryModule } from '@app/modules/token_factory/module';
import { burnSchema, createTokenSchema as createSchema } from '@app/modules/token_factory/schemas';
import { TokenStore } from '@app/modules/token_factory/stores/token';
import { ModuleConfig } from '@app/modules/token_factory/types';
import { TokenID, createCtx, createSampleTransaction } from '@test/helpers';
import { CommandExecuteContext, Transaction, VerifyStatus, chain, codec, db } from 'klayr-sdk';

describe('BurnCommand', () => {
	const initConfig = {
		minAmountToBurn: BigInt(1000),
		chainID: Buffer.from('01234567'),
	};
	const tokenID = new TokenID(1).toBuffer();

	let burnCommand: BurnCommand;
	let createCommand: CreateTokenCommand;
	let stateStore: any;
	let tokenStore: TokenStore;

	const mockMint = jest.fn();
	const mockInitialize = jest.fn();
	const mockBurn = jest.fn();
	const getAvailableBalance = jest.fn();
	const mockPayFee = jest.fn();

	beforeEach(async () => {
		const { minAmountToBurn } = initConfig;
		const tokenFactory = new TokenFactoryModule();
		const tokenFactoryMethod = new TokenFactoryMethod(tokenFactory.stores, tokenFactory.events);
		const internalMethod = new InternalMethod(tokenFactory.stores, tokenFactory.events);
		await internalMethod.init(initConfig.chainID);

		burnCommand = new BurnCommand(tokenFactory.stores, tokenFactory.events);
		burnCommand.addDependencies({
			tokenFactoryMethod,
			tokenMethod: { burn: mockBurn, getAvailableBalance } as any,
		});
		await burnCommand.init({ minAmountToBurn });

		createCommand = new CreateTokenCommand(tokenFactory.stores, tokenFactory.events);
		createCommand.addDependencies({
			internalMethod,
			tokenMethod: { mint: mockMint, initializeToken: mockInitialize },
			feeMethod: { payFee: mockPayFee },
		} as any);
		await createCommand.init(initConfig as ModuleConfig);

		stateStore = new chain.StateStore(new db.InMemoryDatabase());
		tokenStore = tokenFactory.stores.get(TokenStore);
		getAvailableBalance.mockResolvedValue(BigInt(1e4) * BigInt(1e8));
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(burnCommand.name).toEqual('burn');
		});

		it('should have valid schema', () => {
			expect(burnCommand.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			beforeEach(async () => {
				const defaultToken = {
					name: 'The real pepe coin',
					symbol: 'PEPE',
					totalSupply: BigInt(1e4) * BigInt(1e8),
				};
				const defaultValidParams = codec.encode(createSchema, defaultToken);
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const ctx = createCtx<CreateTokenParams>(stateStore, transaction, createSchema, 'execute');

				await expect(
					createCommand.execute(ctx as CommandExecuteContext<CreateTokenParams>),
				).resolves.toBeUndefined();
			});

			it('should throw when `amount` is higher than balance of sender', async () => {
				const paramWithInvalidamount = codec.encode(burnSchema, {
					tokenID,
					amount: BigInt(1e4) * BigInt(1e8) + BigInt(1),
				});
				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidamount, BurnCommand.name),
				);
				const ctx = createCtx<BurnParams>(stateStore, transaction, burnSchema, 'verify');

				const result = await burnCommand.verify(ctx);
				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(
					new Error(
						`Amount can not be lower than ${initConfig.minAmountToBurn} or greater than balance`,
					),
				);
			});

			it('should throw when `amount` is too low', async () => {
				const paramWithInvalidamount = codec.encode(burnSchema, {
					tokenID,
					amount: initConfig.minAmountToBurn - BigInt(1),
				});
				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidamount, BurnCommand.name),
				);
				const ctx = createCtx<BurnParams>(stateStore, transaction, burnSchema, 'verify');

				const result = await burnCommand.verify(ctx);
				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(
					new Error('Amount can not be lower than 1000 or greater than balance'),
				);
			});

			it('should throw when sender is not the token owner / creator', async () => {
				const differentSender = 'ce2e1c458bb616e2717aa48f7392b39b1bc01af7297755db7b6d1db605fa0ef7';
				const validParams = codec.encode(burnSchema, {
					tokenID,
					amount: BigInt(1000) + BigInt(1),
				});
				const transaction = new Transaction(
					createSampleTransaction(validParams, BurnCommand.name, differentSender),
				);
				const ctx = createCtx<BurnParams>(stateStore, transaction, burnSchema, 'verify');

				const result = await burnCommand.verify(ctx);

				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(new Error('Sender is not the token creator'));
			});

			it('should throw correct error when tokenID is invalid', async () => {
				const invalidTokenID = new TokenID(2).toBuffer();
				const paramsWithInvalidTokenID = codec.encode(burnSchema, {
					tokenID: invalidTokenID,
					amount: BigInt(1000) + BigInt(1),
				});
				const transaction = new Transaction(
					createSampleTransaction(paramsWithInvalidTokenID, BurnCommand.name),
				);
				const ctx = createCtx<BurnParams>(stateStore, transaction, burnSchema, 'verify');

				const result = await burnCommand.verify(ctx);

				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(new Error(`Invalid tokenID: ${invalidTokenID}`));
			});

			it('should be OK when params are correct', async () => {
				const validParams = codec.encode(burnSchema, {
					tokenID,
					amount: BigInt(1e2) * BigInt(1e8),
				});
				const transaction = new Transaction(createSampleTransaction(validParams, BurnCommand.name));
				const ctx = createCtx<BurnParams>(stateStore, transaction, burnSchema, 'verify');

				const result = await burnCommand.verify(ctx);
				expect(result.status).toBe(VerifyStatus.OK);
			});
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			beforeEach(async () => {
				const defaultToken = {
					name: 'The real pepe coin',
					symbol: 'PEPE',
					totalSupply: BigInt(1e4) * BigInt(1e8),
				};
				const defaultValidParams = codec.encode(createSchema, defaultToken);
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const ctx = createCtx<CreateTokenParams>(stateStore, transaction, createSchema, 'execute');

				await expect(
					createCommand.execute(ctx as CommandExecuteContext<CreateTokenParams>),
				).resolves.toBeUndefined();
			});

			it('burn function should have been called and updated `totalSupply` ', async () => {
				const validParams = codec.encode(burnSchema, {
					tokenID,
					amount: BigInt(1e2) * BigInt(1e8),
				});

				const transaction = new Transaction(createSampleTransaction(validParams, BurnCommand.name));
				const ctx = createCtx<BurnParams>(stateStore, transaction, burnSchema, 'execute');

				await expect(
					burnCommand.execute(ctx as CommandExecuteContext<BurnParams>),
				).resolves.toBeUndefined();

				const token = await tokenStore.get(ctx, tokenID);

				// initial supply - burned amount
				expect(token.totalSupply).toBe(BigInt(1e4) * BigInt(1e8) - BigInt(1e2) * BigInt(1e8));
				expect(mockBurn).toHaveBeenCalledTimes(1);
			});
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
