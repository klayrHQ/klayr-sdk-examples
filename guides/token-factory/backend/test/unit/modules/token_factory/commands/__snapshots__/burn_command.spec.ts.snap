// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BurnCommand constructor should have valid schema 1`] = `
{
  "$id": "token_factory/burnParams",
  "properties": {
    "amount": {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
    "tokenID": {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
  },
  "required": [
    "tokenID",
    "amount",
  ],
  "title": "Burns tokens that the creator of the token owns",
  "type": "object",
}
`;
